File: lua/testa_test.lua
local testa = require 'testa'

local testTesta = {}

testTesta.correctAddition = testa.is(
    function ()
        return 1 + 2
    end,
    3
)

testTesta.wrongAddition = testa.is(
    function ()
        return 2 + 2
    end,
    3
)

testTesta.correctMultiple = testa.eq(
    function (x, y)
        return x * y
    end,
    function (x, y)
        local r = 0
        for i = 1, y do 
            r = r + x
        end
        return r
    end,
    function (case)
        for i = 1, 5 do
            for j = 1, 5 do
                local ok, msg = case(i, j)
                if not ok then
                    return false, msg
                end
            end
        end
        return true
    end
)

testTesta.wrongMultiple = testa.eq(
    function (x, y)
        return x + y
    end,
    function (x, y)
        local r = 0
        for i = 1, y do 
            r = r + x
        end
        return r
    end,
    function (case)
        for i = 1, 5 do
            for j = 1, 5 do
                local ok, msg = case(i, j)
                if not ok then
                    return false, msg
                end
            end
        end
        return true
    end
)

local function gcd(a, b)
    if a == 0 and b == 0 then
        error("undefined on (0,0)")
    end
    if a == 0 then
        return b
    else
        return gcd(b % a, a)
    end
end

local function gcdtb(case)
    for i = 0, 12 do
        for j = 0, 12 do
            if i ~= 0 or j ~= 0 then
                local ok, msg = case(i, j)
                if not ok then
                    return false, msg
                end
            end
        end
    end
    return true
end

testTesta.correctGcd = testa.verify(
    gcd,
    function (result, i, j)
        if i % result == 0 and j % result == 0 then
            return true
        else
            return false, string.format('%d~=gcd(%d,%d)', result, i, j)
        end
    end,
    gcdtb
)

testTesta.wrongGcd = testa.verify(
    gcd,
    function (result, i, j)
        if i % result ~= 0 or j % result ~= 0 then
            return true
        else
            return false, string.format('%d~=gcd(%d,%d)', result, i, j)
        end
    end,
    gcdtb
)

testa.main({TestTesta = testTesta})

File: erlang/testa_test.erl
%% -*- erlang -*-

%% This file is picked from project testa [https://github.com/TimeExceed/testa.git]
%% Copyright (c) 2017, Taoda (tyf00@aliyun.com)
%% All rights reserved.

%% Redistribution and use in source and binary forms, with or without modification,
%% are permitted provided that the following conditions are met:

%% * Redistributions of source code must retain the above copyright notice, this
%%   list of conditions and the following disclaimer.

%% * Redistributions in binary form must reproduce the above copyright notice, this
%%   list of conditions and the following disclaimer in the documentation and/or
%%   other materials provided with the distribution.

%% * Neither the name of the {organization} nor the names of its
%%   contributors may be used to endorse or promote products derived from
%%   this software without specific prior written permission.

%% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
%% ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
%% WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
%% DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
%% ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
%% (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
%% LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
%% ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
%% (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
%% SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

-module(testa_test).

multipleTb_(_, []) -> ok;
multipleTb_(Case, [H|T]) -> 
    case Case(H) of
        ok -> multipleTb_(Case, T);
        {error, Msg} -> {error, Msg}
    end.
            
multipleTb(Case) ->
    Ins = [{X, Y} || X <- lists:seq(1, 5), Y <- lists:seq(1, 5)],
    multipleTb_(Case, Ins).

gcd(X, Y) ->
    if Y == 0 -> X;
       true -> gcd(Y, X rem Y)
    end.

gcdtb_(_, []) -> ok;
gcdtb_(Case, [H|T]) -> 
    case Case(H) of
        ok -> gcdtb_(Case, T);
        {error, Msg} -> {error, Msg}
    end.

gcdtb(Case) ->
    Ins = [{X, Y} || X <- lists:seq(0, 12), Y <- lists:seq(0, 12), (X /= 0) or (Y /= 0)],
    gcdtb_(Case, Ins).
            
main(Args) ->
    Cases = #{
      "testTesta.correctAddition" => 
          testa:is(
            fun() -> 1 + 2 end,
            3),
      "testTesta.wrongAddition" =>
          testa:is(
           fun() -> 1 + 2 end,
           2),
      "testTesta.correctMultiple" =>
          testa:eq(
            fun({X,Y}) -> X*Y end,
            fun({X,Y}) -> lists:sum([X || _ <- lists:seq(1, Y)]) end,
            fun(X) -> multipleTb(X) end),
      "testTesta.wrongMultiple" =>
          testa:eq(
            fun({X,Y}) -> X+Y end,
            fun({X,Y}) -> lists:sum([X || _ <- lists:seq(1, Y)]) end,
            fun(X) -> multipleTb(X) end),
      "testTesta.correctGcdIs" =>
          testa:is(
            fun({X, Y}) -> gcd(X, Y) end,
            4,
            fun(Case) -> Case({8, 12}) end),
      "testTesta.correctGcd" =>
          testa:verify(
            fun({X,Y}) -> gcd(X, Y) end,
            fun(Result, {X, Y}) ->
                    if (X rem Result == 0) and (Y rem Result == 0) -> ok;
                       true -> {error, io_lib:format("~p/=gcd(~p,~p)", [Result, X, Y])}
                    end
            end,
            fun(X) -> gcdtb(X) end),
      "testTesta.wrongGcd" =>
          testa:verify(
            fun({X,Y}) -> gcd(X, Y) end,
            fun(Result, {X, Y}) ->
                    if (X rem Result /= 0) or (Y rem Result /= 0) -> ok;
                       true -> {error, io_lib:format("~p/=gcd(~p,~p)", [Result, X, Y])}
                    end
            end,
            fun(X) -> gcdtb(X) end)},
    testa:main(Args, Cases).

